service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java

```diff
@@ -0,0 +1,108 @@
+
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.PREVIOUS_UPDATE_PARAMETERS;
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.UPDATE_PARAMETERS;
+
+import bio.terra.cloudres.google.notebooks.AIPlatformNotebooksCow;
+import bio.terra.cloudres.google.notebooks.InstanceName;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.generated.model.ApiGcpAiNotebookUpdateParameters;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.workspace.GcpCloudContextService;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * {@link Step} to update cloud attributes (e.g. metadata) for a ai notebook instance.
+ */
+public class UpdateAiNotebookAttributesStep implements Step {
+  private final ControlledAiNotebookInstanceResource resource;
+  private final CrlService crlService;
+  private final GcpCloudContextService cloudContextService;
+
+  private final Logger logger = LoggerFactory.getLogger(UpdateAiNotebookAttributesStep.class);
+
+  UpdateAiNotebookAttributesStep(
+      ControlledAiNotebookInstanceResource resource,
+      CrlService crlService,
+      GcpCloudContextService gcpCloudContextService
+  ) {
+    this.resource = resource;
+    this.crlService = crlService;
+    this.cloudContextService = gcpCloudContextService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final FlightMap inputMap = context.getInputParameters();
+    final ApiGcpAiNotebookUpdateParameters updateParameters =
+        inputMap.get(UPDATE_PARAMETERS, ApiGcpAiNotebookUpdateParameters.class);
+
+    return updateAiNotebook(updateParameters);
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    final FlightMap workingMap = context.getWorkingMap();
+    final ApiGcpAiNotebookUpdateParameters prevParameters =
+        workingMap.get(PREVIOUS_UPDATE_PARAMETERS, ApiGcpAiNotebookUpdateParameters.class);
+    var projectId =
+        cloudContextService.getRequiredGcpProject(resource.getWorkspaceId());
+    InstanceName instanceName = resource.toInstanceName(projectId);
+    AIPlatformNotebooksCow notebooks = crlService.getAIPlatformNotebooksCow();
+    try {
+      var currentMetadata = notebooks.instances().get(instanceName).execute().getMetadata();
+      for (var entry: currentMetadata.entrySet()) {
+        if (!prevParameters.getMetadata().containsKey(entry.getKey())) {
```

Comment:
iiuc this will delete metadata values for any newly-introduced keys.
Do we also need to reset values for any metadata keys that were modified (i.e. the key is in `prevParameters.getMetadata()` but the value is different)?

Reply:
good catch.